### PR TITLE
fenix: log channels to SQLite database

### PR DESCRIPTION
**Pull request by citrus-it**

### CODE REVIEW

**Review by matt-fidd — DISMISSED** [review]

File: lib/Fenix/Model/IRC.pm [hunk 1]
```diff
@@ -44,6 +45,9 @@ has handlers => sub($self) {
         } keys %{$self->handler}
     ]
 };
+has sqlite   => sub($self) {
+    Mojo::SQLite->new->from_filename(Mojo::File->new($self->datadir, 'irclog.db'));
```
Comment (oetiker) [inline lib/Fenix/Model/IRC.pm:49]: if you want so set any pragmas on the sqllite db than something like this is handy

```suggestion
    my $sql = Mojo::SQLite->new->from_filename(Mojo::File->new($self->datadir, 'irclog.db'));
    
    #$sql->on(connection => sub ($sql, $dbh) {
    #    $dbh->do('PRAGMA foreign_keys = ON;');
    #});
    return $sql
```

**Review by oetiker — APPROVED** [review]

File: lib/Fenix/Model/IRC.pm [hunk 2]
```diff
@@ -181,6 +223,91 @@ sub start($self) {
...
+
+CREATE TABLE nick (
+     nick_id INTEGER PRIMARY KEY AUTOINCREMENT,
+     nick TEXT UNIQUE NOT NULL
+);
+
+CREATE TABLE message (
+     message_id INTEGER PRIMARY KEY AUTOINCREMENT,
+     ts DATETIME NOT NULL,
+     command_id INTEGER,
```
Comment (oetiker) [inline lib/Fenix/Model/IRC.pm:251]:
```suggestion
     command_id INTEGER NOT NULL REFERENCES command(command_id)
```
or can it be NULL ?

File: lib/Fenix/Model/IRC.pm [hunk 3]
```diff
@@ -181,6 +223,91 @@ sub start($self) {
...
+     channel_id INTEGER,
```
Comment (oetiker) [inline lib/Fenix/Model/IRC.pm:252]:
```suggestion
     channel_id INTEGER NOT NULL REFERENCES channel(channel_id),
```

File: lib/Fenix/Model/IRC.pm [hunk 4]
```diff
@@ -181,6 +223,91 @@ sub start($self) {
...
+     nick_id INTEGER,
```
Comment (oetiker) [inline lib/Fenix/Model/IRC.pm:253]:
```suggestion
     nick_id INTEGER NOT NULL REFERENCES nick(nick_id),
```

File: lib/Fenix/Model/IRC.pm [hunk 5]
```diff
@@ -181,6 +223,91 @@ sub start($self) {
...
+     message TEXT,
+     FOREIGN KEY(command_id) REFERENCES command(command_id),
+     FOREIGN KEY(channel_id) REFERENCES channel(channel_id),
+     FOREIGN KEY(nick_id) REFERENCES nick(nick_id)
```
Comment (oetiker) [inline lib/Fenix/Model/IRC.pm:257]:
```suggestion
     message TEXT
```

**Review by oetiker — DISMISSED** [review]
I would suggest adopting a strict naming regime to get unique field names, this is very helpful since that allow to write select statements without using 'as' in most cases ... 

```sql
CREATE TABLE nick (
     nick_id INTEGER PRIMARY KEY AUTOINCREMENT,
     nick_name TEXT UNIQUE NOT NULL
);
CREATE TABLE channel (
     channel_id INTEGER PRIMARY KEY AUTOINCREMENT,
     channel_name TEXT UNIQUE NOT NULL,
     channel_topic TEXT
);
CREATE TABLE message (
     message_id INTEGER PRIMARY KEY AUTOINCREMENT,
     message_ts DATETIME NOT NULL,
     message_command INTEGER NOT NULL REFERENCES command(command_id)
     message_channel INTEGER NOT NULL REFERENCES channel(channel_id),
     message_nick INTEGER NOT NULL REFERENCES nick(nick_id),
     message_text TEXT
);
```

something like this ...

File: lib/Fenix/Model/IRC.pm (hunk 5 above)
Comment (citrus-it) [reply lib/Fenix/Model/IRC.pm:257]: I've changed this to put the attributes alongside the fields, it looks neater that way, thanks.

**Comment by citrus-it** [comment]
> I would suggest adopting a strict naming regime to get unique field names, this is very helpful since that allow to write select statements without using 'as' in most cases ...

I think that's just a matter of preference. I don't like the duplication of the namespace into the elements that causes, and it's as easy to type `message.ts` as it is `message_ts`, and use `as` to remap fields as necessary in the output. For simple ID/value tables, I name the value to match the table. I find that there's a benefit to being able to use `USING` too, making query lines shorter which is easier to read.

In this simple schema, we can get the search results as we want them with something like:
```
                select log.*, rank from log
                join fts_message f using(message_id)
                where f.message match 'fpu OR register*'
                and nick = 'fenix'
                and channel = 'illumos'
                order by rank;
```

**Comment by oetiker** [comment]
Yes I also have a tendency to take shortcuts with 'small' projects ...

**Pull request merged**